### ISSUE TITLE
Support programmatic pairing selection instead of build constraints

### ISSUE THREAD

**Issue by Nik-U** [ConsenSys/gnark#19]
Currently, choosing between BN256, BLS377, and BLS381 is a compile-time decision controlled by build constraints. If none of `bn256`, `bls377`, or `bls381` are provided as build tags, the library is currently unusable.

The current situation is undesirable for several reasons:
- Go libraries should be usable with a simple `go get` and without the need for a `Makefile` or adding custom build tags. Relying on the Go toolchain to handle all of this allows the project to be painlessly used as a (possibly transitive) dependency.
- Users of the library may need to work with a specific pairing. For example, proving statements about a custom twisted Edwards curve requires that the curve's field is the order of G1, and so users that provide such curves and make assumptions about them must also have a guarantee that gnark is using the correct pairing.
- Build constraints in Go are primarily meant for including architecture-specific optimizations (see for example the use in `big` and other uses in the standard library). The infrastructure is built around this intended use. Because of this, gnark is currently difficult to use as a transitive dependency. For example, if a developer writing an application depends on a networking library, that networking library depends on a crypto library, and that crypto library depends on gnark, then the application developer is now responsible for choosing the pairing used by gnark even though they are likely in no position to make an informed choice.
- If Library 1 depends on gnark and requires BLS381, and Library 2 depends on gnark and requires BLS377, then no application can depend on both Library 1 and Library 2.

While most of these issues could be solved by simply making one pairing the default (e.g., BLS381) when the other build tags are not present, that would pose a problem in the future if the default option is shown to have a security flaw, because then the default could not be changed without breaking old users.

The behavior that I would expect as a library user after this enhancement would be:
- Expose all of the different pairing choices in the API and let the user choose between them based on which packages they import.
- In cases where exported functions need to access the underlying pairing, allow the user to pass in the pairing that is in use (either with an interface implemented by other parts of the library or, less desirably, with a constant) or extract that information from the unexported contents of other parameters.
- If a default is provided, then ensure that there is some way to detect that an old default is in use and behave accordingly if the default is changed in the future.
- Export information about the different pairing choices (e.g., the order of the groups). It would be convenient for users requiring a specific pairing to ask the library for the modulus rather than assigning a `big.Int` themselves.

**Comment by gbotrel** [comment]
Fair points, here is the though process we had and the (current) options:

1. Currently, `gnark` uses `bls377` as default, if no build tags is specified. So all the `go` commands should work. The `Makefile` is for users needing a CLI, supporting all curves. 
2. #16 is on our short-term roadmap --> in short, the `ecc` package won't be in  `gnark`, and `gnark` will just import the curve it needs through `go.mod`.
3. If users need to add their own curve (which we don't suspect is a widespread need), that's generated through our `ecc` generation, I suspect the best path would be either to open a PR to `gnark` to add a build tag and a `cs/internal/curve.go`, or fork `gnark` . Is that what you had in mind or do you mean using an existing `gnark` curve code, with a different pairing function? 

We rely on build tags for performance reasons. 
In an ideal world, we have generics, and groth16 algorithms use them . 
Another option, in the `cs` package, is to use interfaces for the curve, the points and the field elements. Problem is: as the circuits scale, we manipulate hundred of millions of these objects and interfaces being pointer, we have (significant) GC overhead and performance impact (memallocs).  At a minimum, we need the prover data structures to be pointer-less (no slice, no interface, no pointers). 
Yet another option, is to duplicate `cs` code through code generation, and have them sit in either different subpackages or different github branches, solving the transitive dependency issue you raised.

It is still unclear to us how to achieve "having curve as a programmatic parameter" to the whole `cs` package, will keep the issue open and see if we can find a solution that's elegant and doesn't affect performance.

**Comment by Nik-U** [comment]
That makes sense. After using gnark for a bit longer I think that explaining my particular use will help to get at the underlying issues.

I am writing a cryptography library that is intended to be used by other developers. I'm using gnark internally to prove and verify some statements, but the developers using my library should not know or care about this. Specifically, I am using the twisted Edwards curves provided by gnark. I am using the group operations and gadgets in `cs/std/{gadget,reference}/algebra/twisted_edwards` for the proofs and public computations.

It makes sense that the Makefile is only for the gnark CLI and not for libraries, which is perfectly reasonable assuming that build tags are required. I was unaware that BLS12-377 is an automatic default because I was using `twisted_edwards.GetEdwardsCurve()` to initialize the `CurveParams` that I use, but while there are curves defined for use with BLS12-381 and BN-256, there is none for BLS12-377. (As an aside, ZEXE does propose parameters for such a curve, but it is not included in gnark: see [IACR 2018/962](https://eprint.iacr.org/2018/962) Figure 16 and [the ZEXE repository](https://github.com/scipr-lab/zexe/tree/master/algebra/src/edwards_bls12/curves) where it is given A = -1 and D = 3021 with cofactor 4 and an allegedly random generator.)

It is the definition of this twisted Edwards curve that depends on the build tags in use, since the `CurveParams` are only valid when G1 has the expected order.

I think that a lot of the problems here can be solved as part of #16. If a `twisted_edwards.GetEdwardsCurve()` function is defined for BLS12-377 using the ZEXE curve, then it will be possible to blindly call that function to get _some_ appropriate curve for the application. However, this still doesn't address the problem that if I were to define my own `CurveParams` and the program was compiled with different build tags, the only indication of anything being wrong would be that proofs would fail.

I can see two clear ways to move forward:

1. Provide a convenient mechanism for callers to determine which curves are in use (i.e., `bls377`, `bls381`, `bn256`, or something else.) While it would be possible to do this by checking the build tags in my own library, this would be brittle if new curves are added to gnark later (since I would need to distinguish between the absence of tags meaning `bls377`, or an unknown tag meaning something else). Something that returns the current `ecc.ID` would work (perhaps it already exists and I just missed it in the documentation, but the only usage GoLand can see is in the gob implementation).

   Given this mechanism, callers would have two choices: they could either ensure that they make no assumptions about the curves in use (most likely the case for most users), or they could use the mechanism to expect a specific setting (probably BLS12-377 to avoid forcing their own users to specify gnark build tags in their application) and panic early if the compilation uses something else.

1. Somehow allow the caller to specify which setting to use at runtime.

The second option was my original suggestion, but it sounds like that might be impossible for performance reasons. I haven't looked at the gnark internals at all, but it sounds like you've already explored this possibility. My initial thought was that you could define an interface with unexported functions that capture the behavior of everything curve-specific in `internal` and use the code generator to produce implementations to pass around at runtime. I understand that this introduces pointer dereferencing overhead, but it's not clear to me why it should necessarily add GC pressure. In any case if benchmarks have shown that the overhead from indirect function calls alone is already too much overhead, then this option is truly a non-starter.

The first option does not solve the dependency problems I mentioned initially, but it does at least allow curve-dependent applications to use gnark in a safe way.

Aside from the lack of a twisted Edwards curve for BLS12-377, the only other issue that I've encountered in my application is that most curve-related functions require a `cs/internal/curve.Element`, which is an internal type. Although I can initialize values with `cs.Element(...)`, my application requires interpreting the value as a `big.Int` in regular form, doing some arithmetic, and converting back to Montgomery form, as well as various other manipulations. I can't see a way to do this other than importing `ecc/bls377/fr` and type converting to `fr.Element`, but this adds an explicit dependency on BLS12-377 when I otherwise might not need one. These scalar operations are also application specific and thus should not be upstreamed to gnark. This particular issue would be solved if `curve.Element` is made public as part of #16 or if `cs_api.go` is expanded to include more operations on `fr.Element`.

**Comment by gbotrel** [comment]
1. You can already do something like:
```
import "github.com/consensys/gnark/cs"
import "github.com/consensys/gnark/ecc"

func init() {
  if cs.CurveID != ecc.BLS377 {
     panic("...")
  }
}
```
Although, that do constrain you to use the default build tag of `gnark`, which is far from ideal. 

2. Problem is "everything curve specific" includes the data structures (proving and verifying key for example) and the field elements that are used throughout the prover (in the constraint solver and the fft for example). Pointer dereferencing is costly for field operations, but memory allocations and [garbage collection indexing](https://hackmd.io/@zkteam/goff#The-cost-of-garbage-collection) is even more so, especially when we scale to very large circuits (>10M constraints), where memory management matters. 

3. So, another idea that I'm looking at, that may be beneficial to this issue (and perf related stuff): generate code for the prover. Technically, usually (as far as we know) users don't need to "dynamically" define circuits. They write the circuit once, run the setup, and then, in application-specific code, prove and verify execution of the circuit. Similarly, we don't see use cases where one would want to **dynamically** change the proving scheme or the underlying curve. What I propose is the following workflow:
a. You write your circuit, with the APIs in `cs_api.go` 
b. From that circuit object, you "export" the groth16 code (ie generate Go code). Which means, in your library, you have an "internal" `go generate` action to generate go code that is typed to a curve (and, not related to dependency but to perf, since we know the size of the circuit, we can better manage memory and do some precomputations for the prover). 
c. the generated code mostly only depends on `ecc/bls381[bn256,...]` -- your library users don't need build tags. Your library need build tags only to re-generate the circuit code.

--> Do you think that would fit your needs?

-----

On the other issue you mentioned, about exposing `cs.Element` . Correct me if I'm wrong: you're talking about the circuit definition phase, correct? That is, when you use `std/reference/algebra` (the `reference` part, not the `gadget` one?), you're force to call `func NewPoint(x, y curve.Element) Point {` or `func (p *Point) ScalarMul(p1 *Point, ecurve CurveParams, scalar curve.Element) *Point {` but handling `curve.Element` type on your side is not .. possible without strong typing, correct?

**Comment by Nik-U** [comment]
Ah, somehow I missed the `cs.CurveID` field. Yes, that at least makes it safe to rely on a particular curve, but does not help with multiple dependencies relying on different curves. The performance issues also make sense; thanks for clarifying.

Indeed, I do not need to dynamically generate circuits for my application. Since defining a new circuit requires a new trusted setup, applications that do this dynamically would need some fairly elaborate infrastructure for dynamic trusted setups that would be eyebrow-raising to say the least. The `go generate` solution would be excellent and would solve all the problems I've raised in this issue (at the cost of excluding that dynamic circuit definition case).

For the `cs.Element` issue, the problem is not just during circuit generation but mainly during circuit usage. I'm defining a circuit that is used to prove statements about curve points. However, these points are the output of the actual computation that I'm doing, and that computation requires doing arithmetic on scalars and performing scalar multiplications on the twisted Edwards curve. I do these computations with the implementation in `std/reference/algebra`. I cannot explicitly use the `curve.Element` type because it is in an `internal` package (attempting to use it produces a compiler error). Using `fr.Element` works, but `fr` refers to different packages based on the build tags.

So in the case that circuit generation is moved into `go generate`, I'd still need some way to work with scalars and points on the twisted Edwards curve that I'm expecting for my circuit.

**Comment by gbotrel** [comment]
Sorry this took so long to close, this is addressed in v0.3.X (and merged into `master`)
